python/cudf_polars/cudf_polars/dsl/expr.py

```diff
@@ -0,0 +1,1023 @@
+
+        Parameter
+        ---------
+        other
+            object to compare to
+
+        Returns
+        -------
+        True if the two expressions are equal, false otherwise.
+        """
+        if type(self) is not type(other):
+            return False
+        return self._ctor_arguments(self.children) == other._ctor_arguments(
+            other.children
+        )
+
+    def __eq__(self, other):
+        """Equality of expressions."""
+        if type(self) != type(other) or hash(self) != hash(other):
+            return False
+        else:
+            return self.is_equal(other)
+
+    def __ne__(self, other):
+        """Inequality of expressions."""
+        return not self.__eq__(other)
+
+    def __repr__(self):
+        """String representation of an expression with caching."""
+        try:
+            return self.repr_value
+        except AttributeError:
+            args = ", ".join(f"{arg!r}" for arg in self._ctor_arguments(self.children))
+            self.repr_value = f"{type(self).__name__}({args})"
+            return self.repr_value
+
+    # TODO: return type is a lie for Literal
+    def evaluate(
+        self,
+        df: DataFrame,
+        *,
+        context: ExecutionContext = ExecutionContext.FRAME,
+        mapping: dict[Expr, Column] | None = None,
+    ) -> Column:
+        """
+        Evaluate this expression given a dataframe for context.
+
+        Parameters
+        ----------
+        df
+            DataFrame that will provide columns.
+        context
+            What context are we performing this evaluation in?
+        mapping
+            Substitution mapping from expressions to Columns, used to
+            override the evaluation of a given expression if we're
+            performing a simple rewritten evaluation.
+
+        Returns
+        -------
+        Column representing the evaluation of the expression (or maybe
+        a scalar, annoying!).
+
+        Raises
+        ------
+        NotImplementedError if we couldn't evaluate the expression.
+        Ideally all these are returned during translation to the IR,
+        but for now we are not perfect.
+        """
+        raise NotImplementedError
+
+    def collect_agg(self, *, depth: int) -> AggInfo:
+        """
+        Collect information about aggregations in groupbys.
+
+        Parameters
+        ----------
+        depth
+            The depth of aggregating (reduction or sampling)
+            expressions we are currently at.
+
+        Returns
+        -------
+        Aggregation info describing the expression to aggregate in the
+        groupby.
+
+        Raises
+        ------
+        NotImplementedError if we can't currently perform the
+        aggregation request (for example nested aggregations like
+        ``a.max().min()``).
+        """
+        raise NotImplementedError
+
+
+def with_mapping(fn):
+    """Decorate a callback that takes an expression mapping to use it."""
+
+    def _(
```

Comment:
AFAICT the `evaluate` method in every subclass of `Expr` uses this. How about defining `evaluate` like this function in `Expr`, then having an internal helper `do_evaluate` or so that is the abstract method implemented by every child class?